src/core/server/uuid/resolve_uuid.ts

```diff
@@ -0,0 +1,99 @@
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import uuid from 'uuid';
+import { promisify } from 'util';
+import * as fs from 'fs';
+import { join } from 'path';
+import { take } from 'rxjs/operators';
+import { IConfigService } from '../config';
+import { PathConfigType, config as pathConfigDef } from '../path';
+import { HttpConfigType, config as httpConfigDef } from '../http';
+import { Logger } from '../logging';
+
+const FILE_ENCODING = 'utf8';
+const FILE_NAME = 'uuid';
+
+const readFile = promisify(fs.readFile);
+const writeFile = promisify(fs.writeFile);
+
+export async function resolveInstanceUuid(
+  configService: IConfigService,
+  logger: Logger
+): Promise<string> {
+  const [pathConfig, serverConfig] = await Promise.all([
+    configService
+      .atPath<PathConfigType>(pathConfigDef.path)
+      .pipe(take(1))
+      .toPromise(),
+    configService
+      .atPath<HttpConfigType>(httpConfigDef.path)
+      .pipe(take(1))
+      .toPromise(),
+  ]);
+
+  const uuidFilePath = join(pathConfig.data, FILE_NAME);
+
+  const uuidFromFile = await readUuidFromFile(uuidFilePath);
+  const uuidFromConfig = serverConfig.uuid;
+
+  if (uuidFromConfig) {
+    if (uuidFromConfig === uuidFromFile) {
+      // uuid matches, nothing to do
+      logger.debug(`Kibana instance UUID: ${uuidFromConfig}`);
+      return uuidFromConfig;
+    } else {
+      // uuid in file don't match, or file was not present, we need to write it.
+      if (uuidFromFile === undefined) {
+        logger.debug(`Setting new Kibana instance UUID: ${uuidFromConfig}`);
+      } else {
+        logger.debug(`Updating Kibana instance UUID to: ${uuidFromConfig} (was: ${uuidFromFile})`);
+      }
+      await writeUuidToFile(uuidFilePath, uuidFromConfig);
+      return uuidFromConfig;
+    }
+  }
+  if (uuidFromFile === undefined) {
+    const newUuid = uuid.v4();
+    // no uuid either in config or file, we need to generate and write it.
+    logger.debug(`Setting new Kibana instance UUID: ${newUuid}`);
+    await writeUuidToFile(uuidFilePath, newUuid);
+    return newUuid;
+  }
+
+  logger.debug(`Resuming persistent Kibana instance UUID: ${uuidFromFile}`);
+  return uuidFromFile;
+}
+
+async function readUuidFromFile(filepath: string): Promise<string | undefined> {
+  try {
+    const content = await readFile(filepath);
+    return content.toString(FILE_ENCODING);
+  } catch (e) {
+    if (e.code === 'ENOENT') {
+      // non-existent uuid file is ok, we will create it.
+      return undefined;
+    }
+    throw e;
```

Comment:
It's possible that we get a `EACCESS` error here and in `writeUuidToFile` we might also get `EISDIR`. I think it would be worth catching these errors and throwing a more useful error to users, maybe something like:

> Unable to write Kibana UUID file, please check the uuid.server configuration value in kibana.yml and ensure Kibana has sufficient permissions to read / write to this file.

Reply:
I think it would also be worth testing these error conditions

Reply:
Explicit error message is indeed better.

Reply:
Explicit error message is indeed better.